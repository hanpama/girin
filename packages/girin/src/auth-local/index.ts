import * as argon2 from 'argon2';
import * as jwt from 'jsonwebtoken';

import { Model, field } from '../mongodb';
import { ObjectID } from 'bson';
import AuthLocalModule from './module';
import { globalEnvironment } from '@girin/environment';


function getPasswordSalt() {
  return globalEnvironment.get(AuthLocalModule).PASSWORD_SALT;
}
function getJWTSecretKey() {
  return globalEnvironment.get(AuthLocalModule).JWT_SECRET_KEY;
}

export type BaseUserClass<T extends BaseUser> = typeof BaseUser & {
  new(source: any): T;
}

export class BaseUser extends Model {

  @field()
  protected hashedPassword: string;

  public static async createUserWithPassword(source: { [fieldName: string]: any }, password: string) {

    const hashedPassword = await argon2.hash(password + getPasswordSalt());
    const user = await this.insert({ hashedPassword, ...source });
    return user;
  }

  public authenticate(password: string) {
    return argon2.verify(this.hashedPassword, password + getPasswordSalt());
  }

  public async changePassword(newPassword: string) {
    const hashedPassword = await argon2.hash(newPassword + getPasswordSalt());
    this.hashedPassword = hashedPassword;
    await this.$update({ $set: { hashedPassword } });

    return this;
  }

  public async encodeToken(password: string) {
    const authenticated = await this.authenticate(password);
    if (!authenticated) {
      throw new Error('Authentication Error');
    }
    const { hashedPassword, ...rest } = this.$source;
    return jwt.sign(rest, getJWTSecretKey());
  }

  public static decodeToken<T extends BaseUser>(this: BaseUserClass<T>, token: string): T {
    const result: any = jwt.verify(token, getJWTSecretKey());
    console.log(result);
    const user = new this(result);
    user._id = new ObjectID(result._id);
    return user;
  }
}
