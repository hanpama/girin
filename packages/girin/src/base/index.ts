export * from './scalar';
export * from './Definition';
export * from './MetadataStorage';
export * from './TypeExpression';
export * from './structure';
