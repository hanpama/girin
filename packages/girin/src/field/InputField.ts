import { GraphQLInputType } from "graphql";
import { TypeExpression, MetadataStorage } from "..";


export interface InputFieldConfig {
  type: TypeExpression;
  defaultName: string;
  defaultValue?: any;
  description?: string;
  directives?: any;
}

export class InputField<TConfig extends InputFieldConfig = InputFieldConfig> {
  constructor(protected config: TConfig) { }

  public get defaultName() { return this.config.defaultName; }
  public get defaultValue() { return this.config.defaultValue; }
  public get description() { return this.config.description; }
  public get directives() { return this.config.directives; }

  public buildType(storage: MetadataStorage, targetClass?: Function): GraphQLInputType {
    return this.config.type.buildTypeInstance(storage, targetClass) as GraphQLInputType;
  }
}
