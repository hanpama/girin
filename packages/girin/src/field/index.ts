export * from './Field';
export * from './InputField';
// export * from './mount';
