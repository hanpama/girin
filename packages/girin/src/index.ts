export { StringScalar, BooleanScalar, FloatScalar, IntScalar, IDScalar } from './builtins/scalar';
export { List, NonNull } from './type-expression/structure';
export * from './base';
export * from './metadata';
export * from './sdl';
export * from './getGraphQLType';
export * from './field';
export * from './type-expression';
