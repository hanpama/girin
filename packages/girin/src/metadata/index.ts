export * from './InputType';
export * from './ObjectType';
export * from './ScalarType';
export * from './InterfaceType';
