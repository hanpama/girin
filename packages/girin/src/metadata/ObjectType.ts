import { GraphQLObjectType, GraphQLFieldConfigMap, GraphQLInterfaceType, GraphQLFieldConfig } from "graphql";

import { Definition, DefinitionConfig } from "../base/Definition";
import { MetadataStorage, FieldReferenceEntry } from "../base/MetadataStorage";
import { TypeExpression } from "../type-expression/TypeExpression";
import { ASTParser } from "../sdl/ast";


export interface ObjectTypeConfig extends DefinitionConfig {
  description?: string;
  interfaces?: TypeExpression[];
}

/**
 * Metadata type for ObjectType
 */
export class ObjectType<TConfig extends ObjectTypeConfig = ObjectTypeConfig> extends Definition<TConfig> {

  protected static decorate(astParser: ASTParser | undefined, storage: MetadataStorage, linkedClass: Function) {
    super.decorate(astParser, storage, linkedClass);

    if (astParser) {
      astParser.objectTypeMetadataConfigs.forEach(config => {
        storage.register(new this(config), linkedClass);
      });
    } else {
      storage.register(new this({ typeName: linkedClass.name }), linkedClass);
    }
  }

  public buildFieldConfig(storage: MetadataStorage, targetClass: Function, entry: FieldReferenceEntry): GraphQLFieldConfig<any, any> {
    const { description, deprecationReason } = entry.field;

    return {
      type: entry.field.buildType(storage, targetClass),
      args: entry.field.buildArgs(storage, targetClass),
      resolve: entry.resolver,
      description,
      deprecationReason,
    };
  }

  public buildFieldConfigMap(storage: MetadataStorage, targetClass: Function): GraphQLFieldConfigMap<any, any> {
    const refs = storage.queryFieldReferences(targetClass);
    return (
      refs.reduce((results, entry) => {
        const name = entry.mountName;

        results[name] = this.buildFieldConfig(storage, targetClass, entry);
        return results;
      }, {} as GraphQLFieldConfigMap<any, any>)
    );
  }

  /**
   * Build GraphQLObjectType instance from metadata.
   */
  public buildTypeInstance(storage: MetadataStorage, targetClass: Function): GraphQLObjectType {
    const name = this.typeName || targetClass.constructor.name;
    const fields = this.buildFieldConfigMap.bind(this, storage, targetClass);
    const interfaces = this.findInterfaces(storage, targetClass);
    const description = this.description;
    const isTypeOf = this.buildIsTypeOf(storage, targetClass);
    return new GraphQLObjectType({ name, fields, interfaces, description, isTypeOf });
  }

  public findInterfaces(storage: MetadataStorage, targetClass: Function): GraphQLInterfaceType[] | undefined {
    const { interfaces } = this.config;
    return interfaces && interfaces.map(i => (
      i.buildTypeInstance(storage, targetClass) as GraphQLInterfaceType)
    );
  }

  public buildIsTypeOf(storage: MetadataStorage, targetClass: Function) {
    return (source: any) => (source instanceof targetClass);
  }
}
