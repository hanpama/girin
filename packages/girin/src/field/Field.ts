import { GraphQLFieldConfigArgumentMap, GraphQLFieldConfig, GraphQLOutputType, GraphQLFieldResolver } from "graphql";
import { TypeExpression, TypeArg } from "../type-expression/TypeExpression";
import { MetadataStorage } from "../base/MetadataStorage";
import { InputFieldReference } from "./InputField";


export interface FieldProps {
  description?: string;
  deprecationReason?: string;
  directives?: any;
  resolve?: Function;
}

export class FieldReference {
  constructor(
    public name: string,
    public field: Field,
    public props: FieldProps,
  ) { }
}

export interface FieldBuilder {
  output: TypeArg | TypeExpression;
  args: InputFieldReference[];
  buildResolver?(storage: MetadataStorage): GraphQLFieldResolver<any, any> | undefined;
  buildArgs?(storage: MetadataStorage): GraphQLFieldConfigArgumentMap;
  buildConfig?(storage: MetadataStorage): GraphQLFieldConfig<any, any>
}

export class Field implements FieldBuilder {

  constructor(builder?: FieldBuilder) {
    if (builder) { Object.assign(this, builder); }
  }

  public output: TypeExpression;
  public args: InputFieldReference[];

  public mountAs(fieldName: string, props?: FieldProps) {
    return new FieldReference(fieldName, this, props || {});
  }

  public buildResolver(storage: MetadataStorage): GraphQLFieldResolver<any, any> | undefined  {
    return;
  }

  public buildArgs(storage: MetadataStorage): GraphQLFieldConfigArgumentMap {
    const { args } = this;
    return args.reduce((args, ref) => {
      args[ref.name] = Object.assign(ref.field.buildConfig(storage), ref.props);
      return args;
    }, {} as GraphQLFieldConfigArgumentMap);
  }

  public buildConfig(storage: MetadataStorage): GraphQLFieldConfig<any, any> {
    const { output } = this;

    return {
      type: output.buildTypeInstance(storage) as GraphQLOutputType,
      args: this.buildArgs(storage),
      resolve: this.buildResolver(storage),
    };
  }
}
