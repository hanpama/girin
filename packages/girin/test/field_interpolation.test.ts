import { gql, getGraphQLType, List, StringScalar } from "../src";
import { Field } from "../src/field/Field";
import { GraphQLObjectType, GraphQLSchema, graphql, printSchema } from "graphql";
import { InputField } from "../src/field/InputField";
import { ObjectType } from "../src/metadata/ObjectType";


const tagListField = new Field({ output: List.of(StringScalar), args: [] });

@ObjectType.define(gql`
  type Room {
    name: String!
    ${tagListField.mountAs('tags')}
  }
`)
class Room {
  name: string;
  tags: string[];
}

class RoomsField extends Field {
  output = List.of(Room)
  args = [
    { name: 'count', field: new InputField('Int'), props: {} }
  ]
  buildResolver() {
    return (source: string, args: { count: number }) => {
      const rooms: Room[] = [];
      for(let i = 0; i < args.count; i ++) {
        const room = new Room();
        room.name = `My room ${i}`;
        room.tags = [String(i)];
        rooms.push(room)
      }
      return rooms;
    }
  }
}

@ObjectType.define(gql`
  type Host {
    name: String!
    ${(new RoomsField()).mountAs('rooms')}
  }
`)
class Host {
  name: string;
  rooms: Room[];
}

@ObjectType.define(gql`
  type Query {
    host: ${Host}
  }
`)
class Query {
  static host() {
    return new Host();
  }
}

describe('Fields', () => {
  it('should append fields to GraphQLType created', () => {
    const roomObjectType: GraphQLObjectType = getGraphQLType(Room);

    const fields = roomObjectType.getFields();
    expect(fields.tags).not.toBeUndefined();
    expect(fields.tags.type.toString()).toBe('[String]');
  });

  it('should create and execute schema as expected', async () => {
    const schema = new GraphQLSchema({
      query: getGraphQLType(Query),
    });

    expect(printSchema(schema)).toMatchSnapshot();

    const results = await graphql({ schema, source: `
      query {
        host {
          rooms(count: 5) {
            name
            tags
          }
        }
      }
    ` });

    expect(results).toEqual({
      "data": {
        "host": {
          "rooms": [
            { "name": "My room 0",
              "tags": [ "0" ] },
            { "name": "My room 1",
              "tags": [ "1" ] },
            { "name": "My room 2",
              "tags": [ "2" ] },
            { "name": "My room 3",
              "tags": [ "3" ] },
            { "name": "My room 4",
              "tags": [ "4" ] },
          ],
        },
      },
    });
  })
});
