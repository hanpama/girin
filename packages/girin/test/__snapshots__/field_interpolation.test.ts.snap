// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fields should create and execute schema as expected 1`] = `
"type Host {
  name: String!
  rooms(count: Int): [Room]
}

type Query {
  host: Host
}

type Room {
  name: String!
  tags: [String]
}
"
`;
